pkg/provisioning/import.go

```diff
@@ -0,0 +1,712 @@
+// -- CREATE ACTIONS --
+// --------------------
+
+type createPipelineAction struct {
+	cfg config.Pipeline
+
+	pipelineService PipelineService
+}
+
+func (a createPipelineAction) String() string {
+	return fmt.Sprintf("create pipeline with ID %v", a.cfg.ID)
+}
+func (a createPipelineAction) Do(ctx context.Context) error {
+	_, err := a.pipelineService.Create(ctx, a.cfg.ID, pipeline.Config{
+		Name:        a.cfg.Name,
+		Description: a.cfg.Description,
+	}, pipeline.ProvisionTypeConfig)
+	if err != nil {
+		return cerrors.Errorf("failed to create pipeline: %w", err)
+	}
+	_, err = a.pipelineService.UpdateDLQ(ctx, a.cfg.ID, pipeline.DLQ{
+		Plugin:              a.cfg.DLQ.Plugin,
+		Settings:            a.cfg.DLQ.Settings,
+		WindowSize:          *a.cfg.DLQ.WindowSize,
+		WindowNackThreshold: *a.cfg.DLQ.WindowNackThreshold,
+	})
+	if err != nil {
+		return cerrors.Errorf("failed to update pipeline DLQ: %w", err)
+	}
+
+	// add connector IDs
+	for _, conn := range a.cfg.Connectors {
+		_, err = a.pipelineService.AddConnector(ctx, a.cfg.ID, conn.ID)
+		if err != nil {
+			return cerrors.Errorf("failed to add connector %v: %w", conn.ID, err)
+		}
+	}
+	// add processor IDs
+	for _, proc := range a.cfg.Processors {
+		_, err = a.pipelineService.AddProcessor(ctx, a.cfg.ID, proc.ID)
+		if err != nil {
+			return cerrors.Errorf("failed to add processor %v: %w", proc.ID, err)
+		}
+	}
+
+	return nil
+}
+func (a createPipelineAction) Rollback(ctx context.Context) error {
+	err := a.pipelineService.Delete(ctx, a.cfg.ID)
+	if !cerrors.Is(err, pipeline.ErrInstanceNotFound) {
+		// ignore instance not found errors, this means the action failed to
+		// create the pipeline in the first place
+		return err
+	}
+	return nil
+}
+
+type createConnectorAction struct {
+	cfg        config.Connector
+	pipelineID string
+
+	connectorService ConnectorService
+	pluginService    PluginService
+}
+
+func (a createConnectorAction) String() string {
+	return fmt.Sprintf("create connector with ID %v", a.cfg.ID)
+}
+func (a createConnectorAction) Do(ctx context.Context) error {
+	_, err := a.connectorService.Create(
+		ctx,
+		a.cfg.ID,
+		a.connectorType(a.cfg.Type),
+		a.cfg.Plugin,
+		a.pipelineID,
+		connector.Config{
+			Name:     a.cfg.Name,
+			Settings: a.cfg.Settings,
+		},
+		connector.ProvisionTypeConfig,
+	)
+	if err != nil {
+		return cerrors.Errorf("failed to create connector: %w", err)
+	}
+
+	// add processor IDs
+	for _, proc := range a.cfg.Processors {
+		_, err = a.connectorService.AddProcessor(ctx, a.cfg.ID, proc.ID)
+		if err != nil {
+			return cerrors.Errorf("failed to add processor %v: %w", proc.ID, err)
+		}
+	}
+
+	return nil
+}
+func (a createConnectorAction) Rollback(ctx context.Context) error {
+	err := a.connectorService.Delete(ctx, a.cfg.ID, a.pluginService)
+	if !cerrors.Is(err, connector.ErrInstanceNotFound) {
+		// ignore instance not found errors, this means the action failed to
+		// create the connector in the first place
```

Comment:
+1 for the comment! At least to me, it's more useful to have them before the `if`, and not within the `if` block, because at this point we're not ignoring ErrInstanceNotFound.